File: Leetcode/leetcode63.cpp
class Solution {
public:
    const int inf = 2e9;
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int row = obstacleGrid.size();
        int col = obstacleGrid[0].size();
        vector <vector <int> > dp (row, vector <int> (col, 0));
        dp[0][0] = 1;
        for (int i = 0; i < row; i++) {
            if (obstacleGrid[i][0] == 1) break;
            else dp[i][0] = 1;
        }
        for (int j = 0; j < col; j++) {
            if (obstacleGrid[0][j] == 1) break;
            else dp[0][j] = 1;
        }
        for (int i = 1; i < row; i++) {
            for (int j = 1; j < col; j++) {
                if (obstacleGrid[i][j] == 1) dp[i][j] = 0;
                else dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return obstacleGrid[row - 1][col - 1] == 0 ? dp[row - 1][col - 1] : 0;
    }
};

File: Leetcode/leetcode63.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        row, col = len(obstacleGrid), len(obstacleGrid[0])
        dp = [[int(0) for i in range(col)] for j in range(row)]

        # for row in dp:
        #     print(' '.join([str(val) for val in row]))

        for i in range(row):
            if obstacleGrid[i][0] == 0:
                dp[i][0] = 1
            else:
                break

        for j in range(col):
            if obstacleGrid[0][j] == 0:
                dp[0][j] = 1
            else:
                break

        for i in range(1, row):
            for j in range(1, col):
                if obstacleGrid[i][j] == 0:
                    dp[i][j] = dp[i - 1][j] + dp[i][j - 1]

        return 0 if obstacleGrid[row - 1][col - 1] == 1 else dp[row - 1][col - 1]